### DISCUSSION TITLE
ValueError: slice step cannot be zero

### DISCUSSION THREAD

**Discussion by alienx-zero** [dipy/dipy#2743] [Tractography / Tractometry]
I try to extract the connectivity in these sentence:
`M, grouping = utils.connectivity_matrix(cc_streamlines, affine, parcellation.astype(int),
                                        return_mapping=True,
                                        mapping_as_streamlines=True)`
However, it rise error:

```python
Traceback (most recent call last):
  File "F:\OPT\research\fMRI\utils_for_all\structural_connectivity_extract.py", line 152, in <module>
    M, grouping = utils.connectivity_matrix(cc_streamlines, np.eye(4), parcellation.astype(int),

  File "d:\Users\99488\anaconda3\envs\torch1.10\lib\site-packages\dipy\tracking\utils.py", line 220, in connectivity_matrix
    endpoints = [sl[0::len(sl)-1] for sl in streamlines]

  File "d:\Users\99488\anaconda3\envs\torch1.10\lib\site-packages\dipy\tracking\utils.py", line 220, in <listcomp>
    endpoints = [sl[0::len(sl)-1] for sl in streamlines]

ValueError: slice step cannot be zero
```

I debug these issue and figure out that for some streamline only one list with 3  value. Always the streamline should be a batch of list of 3 value, right? Any idea how to solve it

**Comment by skoudoro** [comment]
Hi @alienx-zero,

it will be very hard or almost impossible to help if we do not know how you obtain and select your streamlines. 
In this specific case, what is `cc_streamlines`?  how do you obtain it? 

PS: Converting this issue to a discussion since it more an help requested than an issue in DIPY

**Comment by alienx-zero** [comment]
Here is the whole code:

    subject = 'sub-CU0052CUMR1R1'
    save_path = r'H:\PHD\learning\research\dataset\EMBARC_dti\dti_feature'
    hardi_fname = r'H:\PHD\learning\research\dataset\EMBARC_dti\qsiprep\{}\dwi\{}_space-T1w_desc-preproc_dwi.nii.gz'.format(subject, subject)
    hardi_bval_fname = r'H:\PHD\learning\research\dataset\EMBARC_dti\qsiprep\{}\dwi\{}_space-T1w_desc-preproc_dwi.bval'.format(subject, subject)
    hardi_bvec_fname = r'H:\PHD\learning\research\dataset\EMBARC_dti\qsiprep\{}\dwi\{}_space-T1w_desc-preproc_dwi.bvec'.format(subject, subject)
    parcellation_fname = os.path.join(r'F:\OPT\research\fMRI\utils_for_all\parcellation',
                  'SchaeferYeo2018/MNI/Schaefer2018_100Parcels_7Networks_order_plusSubcort_FSLMNI152_2mm.nii')
    label_fname = r'H:\PHD\learning\research\dataset\EMBARC_dti\qsiprep\{}\anat\{}_space-MNI152NLin2009cAsym_label-WM_probseg.nii.gz'.format(subject, subject)
    t1_fname = r'H:\PHD\learning\research\dataset\EMBARC_dti\qsiprep\{}\anat\{}_desc-preproc_T1w.nii.gz'.format(subject, subject)
    save_files_subs = os.path.join(save_path, subject)
    if os.path.exists(save_files_subs) == False:
        os.makedirs(save_files_subs)
    data, affine, hardi_img = load_nifti(hardi_fname, return_img=True)
    t1_data = load_nifti_data(t1_fname)

    labels_resample = resample_img(
        label_fname,
        target_affine=affine,
        target_shape=data.shape[:3],
        interpolation="nearest",
        copy=True)
    labels = get_data(labels_resample)
    # plt.imshow(labels2[90,:,:])
    # labels[labels<0] = 0
    # label2 = nib.Nifti1Image(labels, affine=affine)
    # fa_image = nib.Nifti1Image(labels[labels>0], affine=affine)

    # plot_epi(label_fname)
    # plot_anat(label2, cut_coords = [0, -17, 16])
    parcellation_resample = resample_img(
        parcellation_fname,
        target_affine=affine,
        target_shape=data.shape[:3],
        interpolation="nearest",
        copy=True)
    parcellation = get_data(parcellation_resample)
    
    bvals, bvecs = read_bvals_bvecs(hardi_bval_fname, hardi_bvec_fname)
    gtab = gradient_table(bvals, bvecs)
    
    white_matter = labels>=0
    csamodel = shm.CsaOdfModel(gtab, 6)
    csapeaks = peaks.peaks_from_model(model=csamodel, data=data, sphere=peaks.default_sphere,
                                      relative_peak_threshold=.8, min_separation_angle=45,
                                      mask=white_matter, parallel=True, num_processes=6)
    
    # affine = np.eye(4)
    seeds = utils.seeds_from_mask(white_matter, np.eye(4), density=1)
    stopping_criterion = BinaryStoppingCriterion(white_matter)
    
    streamline_generator = LocalTracking(csapeaks, stopping_criterion, seeds,
                                          affine=np.eye(4), step_size=0.5)
    streamlines = Streamlines(streamline_generator)
    
    # streamlines = load_vtk_streamlines(r'C:\Users\99488\OneDrive\cluster_code\code_file\code\fmri_process\streamlines.vtk')

    cc_slice = (labels !=0)&(parcellation != 0)
    cc_streamlines = utils.target(streamlines, np.eye(4), cc_slice)
    cc_streamlines = Streamlines(cc_streamlines)
    
    # cc_slice = (labels >= 0)
    # cc_streamlines = utils.target(streamlines, np.eye(4), cc_slice)
    # cc_streamlines = Streamlines(cc_streamlines)

    # # endpoints = []
    # # for sl in streamlines:
    # #     if sl.shape[0] == 1:
    # #         endpoints.append(sl[0])
    # #     else:
    # #         endpoints.append(sl[0::len(sl)-1])
            
    # # # Enables/disables interactive visualization
    interactive = True
    
    # # Make display objects
    # color = cmap.line_colors(cc_streamlines)
    # cc_streamlines_actor = actor.line(cc_streamlines,
    #                                   cmap.line_colors(cc_streamlines))
    # cc_ROI_actor = actor.contour_from_roi(cc_slice, color=(1., 1., 0.),
    #                                       opacity=0.5)

    # vol_actor = actor.slicer(t1_data)
    
    # # vol_actor.display(x=40)
    # vol_actor2 = vol_actor.copy()
    # # vol_actor2.display(z=35)
    
    # # Add display objects to canvas
    # scene = window.Scene()
    # scene.add(cc_streamlines_actor)
    # scene.add(cc_ROI_actor)
    # scene.set_camera(position=[-1, 0, 0], focal_point=[0, 0, 0], view_up=[0, 0, 1])
    # # window.record(scene, n_frames=1, out_path=os.path.join(save_files_subs, 'streamlines.png'),
    # #               size=(800, 800))
    # if interactive:
    #     window.show(scene)
    # # save_vtk_streamlines(streamlines, os.path.join(save_files_subs, 'streamlines.vtk'))
    
    M, grouping = utils.connectivity_matrix(cc_streamlines, np.eye(4), parcellation.astype(int),
                                        return_mapping=True,
                                        mapping_as_streamlines=True)

**Comment by alienx-zero** [comment]
If you want, I also can share you this subject's dMRI files

**Comment by alienx-zero** [comment]
@skoudoro, here is the data link https://drive.google.com/drive/folders/1S2BIgXMwkkQeZhaHAsQF4QRyh9uogu6U?usp=sharing, if you need them. The link would be effective in one week

**Comment by alienx-zero** [comment]
Any useful idea or suggestion? if no, I have to switch to another software
